src/main/java/org/jenkinsci/plugins/workflow/cps/CpsVmExecutorService.java

```diff
@@ -105,9 +107,42 @@ private ThreadContext setUp() {
             assert cpsThreadGroup.getExecution().getShell().getClassLoader() != null;
             t.setContextClassLoader(cpsThreadGroup.getExecution().getShell().getClassLoader());
         }
+        CpsCallableInvocation.registerMismatchHandler(this::handleMismatch);
         return context;
     }
 
+    private void handleMismatch(Object expectedReceiver, String expectedMethodName, Object actualReceiver, String actualMethodName) {
+        String mismatchMessage = mismatchMessage(className(expectedReceiver), expectedMethodName, className(actualReceiver), actualMethodName);
+        if (FAIL_ON_MISMATCH) {
+            throw new IllegalStateException(mismatchMessage);
+        } else {
+            try {
+                cpsThreadGroup.getExecution().getOwner().getListener().getLogger().println(mismatchMessage);
+            } catch (IOException x) {
+                LOGGER.log(Level.FINE, null, x);
+            }
+        }
+    }
+
+    private static @CheckForNull String className(@CheckForNull Object receiver) {
+        return receiver == null ? null : receiver.getClass().getName();
+    }
+
+    static boolean FAIL_ON_MISMATCH = Main.isUnitTest;
```

Comment:
Making false positives be fatal makes it much easier to detect mistakes here and in PCT. But we would rather have this be nonfatal in production, since there are sure to be some false positives in exotic situations not yet covered by tests. (As well as some false negatives, but this is a best effort after all.)